src/index.ts

```diff
@@ -11,22 +11,25 @@
  * or implied. See the License for the specific language governing permissions and limitations
  * under the License.
  */
-/* eslint-disable no-var,vars-on-top */
-// eslint-disable-next-line prefer-destructuring
-var createContext = require('react').createContext;
+import React, { type Context } from "react";
 
-var contexts = {};
+// eslint-disable-next-line @typescript-eslint/no-explicit-any -- Allow any type for contexts
+const contexts: { [key: string]: Context<any> | undefined } = {};
 
-module.exports = function createSharedReactContext(defaultValue, key) {
+function createSharedReactContext<T>(defaultValue: T, key: string): Context<T> {
   if (!key) {
-    // eslint-disable-next-line no-console
-    console.warn('Second parameter in createSharedReactContext was not set, defaulting to React.createContext');
-    return createContext(defaultValue);
+    // eslint-disable-next-line no-console -- Warn users if they don't provide a key
+    console.warn(
+      "Second parameter in createSharedReactContext was not set, defaulting to React.createContext",
+    );
+    return React.createContext(defaultValue);
   }
   if (contexts[key]) {
-    return contexts[key];
+    return contexts[key] as Context<T>;
```

Comment:
If the provider part of the application uses

```
createSharedReactContext<{isModalOpen: boolean}>(...)
```

and the consumer uses

```
createSharedReactContext<{modalState: 'open' | 'closed'}>(...)
```

Then the Context produced will be lying to the consumer about the type of the context provided.

To make this code completely 'safe' it would need to always return `Context<unknown>` to force the consumer to validate the context shape explicitly.